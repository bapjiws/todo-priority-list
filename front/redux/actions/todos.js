import * as types from '../actions/types';

export const addTodoSuccess = payload => ({type: types.ADD_TODO_SUCCESS, payload});
export const addTodoFailure = error => ({type: types.ADD_TODO_FAILURE, error});

export const addTodo = todo => {
    return (dispatch, getState, { axiosInstance }) => {
        return axiosInstance({
            method: 'post',
            url: '/addTodo',
            data: { todo }
        })
            .then(response => {
                dispatch(addTodoSuccess(response.data));
                return Promise.resolve();
            })
            .catch(error => {
                dispatch(addTodoFailure(error));
                return Promise.reject();
            });
    }
};

export const loadTodosSuccess = payload => ({type: types.LOAD_TODOS_SUCCESS, payload});
export const loadTodosFailure = error => ({type: types.LOAD_TODOS_FAILURE, error});

export const loadTodos = () => {
    return (dispatch, getState, { axiosInstance }) => {
        return axiosInstance({
            method: 'get',
            url: '/loadTodos',
            params: {
                sort: 'desc'
            }
        })
            .then(response => {
                dispatch(loadTodosSuccess(response.data.map(item => item.fields)));
                return Promise.resolve();
            })
            .catch(error => {
                dispatch(loadTodosFailure(error));
                return Promise.reject();
            });
    }
};
