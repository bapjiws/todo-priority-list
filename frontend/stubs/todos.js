export const currentTodos = [
    {
        priority: 4,
        name: 'eat something',
        description: 'ASAP'
    },
    {
        priority: 3,
        name: 'get some sleep',
        description: 'ASAP, too'
    },
    {
        priority: 1,
        name: 'do some work',
        description: 'meh'
    }
];

export const todoToInsert = {
    priority: 2,
    name: 'do the dishes',
    description: 'not so urgent'
};

export const sortedTodos = [
    {
        priority: 4,
        name: 'eat something',
        description: 'ASAP'
    },
    {
        priority: 3,
        name: 'get some sleep',
        description: 'ASAP, too'
    },
    {
        priority: 2,
        name: 'do the dishes',
        description: 'not so urgent'
    },
    {
        priority: 1,
        name: 'do some work',
        description: 'meh'
    }
];