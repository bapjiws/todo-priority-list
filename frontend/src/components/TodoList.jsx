import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Table } from 'react-bootstrap';

import { loadTodos } from '../../redux/actions/todos';

export class TodoList extends Component {
    componentDidMount() {
        this.props.loadTodos();
    }

    render() {
        return <Table striped bordered condensed className="todos-container">
            <thead>
                <tr>
                    <th>Name</th>
                    <th>Description</th>
                    <th>Priority</th>
                </tr>
            </thead>
            <tbody>
                {
                    this.props.todos.map((todo, idx) => <tr key={idx}>
                        <td>{todo.name}</td>
                        <td>{todo.description}</td>
                        <td>{todo.priority}</td>
                    </tr>)
                }
            </tbody>
        </Table>
    }
}

export default connect(
    state => ({
        todos: state.todos.data
    }),
    { loadTodos }
)(TodoList);