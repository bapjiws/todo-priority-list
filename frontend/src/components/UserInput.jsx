import React, { Component } from 'react';
import { connect } from 'react-redux';
import { FormGroup, FormControl, ControlLabel, Radio, Button } from 'react-bootstrap';

import { addTodo } from '../../redux/actions/todos';

const priorityScale = [1,2,3,4,5,6,7,8,9,10];

export class UserInput extends Component {
    constructor(props) {
        super(props);
        this.state = {
            priority: 1,
            name: '',
            description: ''
        };

        this.handleChange = this.handleChange.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
        this.handleCLick = this.handleCLick.bind(this);
    }

    componentDidMount() {
        document.querySelectorAll('input[type="radio"]')[0].checked = true;
    }

    handleChange(event) {
        const { id, value } = event.target;
        this.setState({
            [id]: id === 'priority' ? +value : value // this.state.priority is now handled by this.handleCLick
        });
    }

    handleSubmit(event) {
        if (this.getNameValidationState() === 'error' || this.getDescriptionValidationState() === 'error') {
            event.preventDefault();
            return;
        }

        event.preventDefault();

        this.props.addTodo(this.state);
        document.querySelectorAll('input[type="radio"]')[0].checked = true;
        this.setState({
            priority: 1,
            name: '',
            description: ''
        });
    }

    handleCLick(event) {
        this.setState({
           priority: +event.target.value
        });
    }

    getNameValidationState() {
        const length = this.state.name.length;
        return length === 0 ? 'error' : 'success';
    }

    getDescriptionValidationState() {
        const length = this.state.description.length;
        return length === 0 ? 'error' : 'success';
    }

    render() {
        return <form className="form-container">
            <FormGroup
                className="form-space-between"
                controlId="priority"
            >
                <ControlLabel>Priority</ControlLabel>
                {' '}
                <span>
                    {
                        priorityScale.map((priority, idx) =>
                            <Radio
                                inline key={idx}
                                name="radioGroup"
                                value={priority}
                                // inputRef={ref => { this.priorities[priority] = ref; }}
                                onClick={this.handleCLick}
                            >
                                {priority}
                            </Radio>)
                    }
                </span>
            </FormGroup>

            <FormGroup
                controlId="name"
                validationState={this.getNameValidationState()}
            >
                <ControlLabel>Name</ControlLabel>
                <FormControl
                    type="text"
                    value={this.state.name}
                    placeholder="Name your task"
                    onChange={this.handleChange}
                />
                <FormControl.Feedback />
            </FormGroup>

            <FormGroup
                controlId="description"
                validationState={this.getDescriptionValidationState()}
            >
                <ControlLabel>Description</ControlLabel>
                <FormControl
                    type="text"
                    value={this.state.description}
                    placeholder="Describe your task"
                    onChange={this.handleChange}
                />
                <FormControl.Feedback />
            </FormGroup>

            <Button type="submit" onClick={this.handleSubmit}>
                Submit
            </Button>
        </form>
    }
}

export default connect(
    null,
    { addTodo }
)(UserInput);