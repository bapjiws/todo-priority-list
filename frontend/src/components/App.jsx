import React from 'react';
import { Provider } from 'react-redux';

import configureStore from '../../redux/configureStore';

import '../../styles/main';

import UserInput from './UserInput';
import TodoList from './TodoList';

const App = () => {
    return <Provider store={configureStore()}>
        <div id="app-container">
            <UserInput/>
            <TodoList/>
        </div>
    </Provider>
};

export default App;