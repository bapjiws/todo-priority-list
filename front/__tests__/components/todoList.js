import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json'; // Documentation: https://github.com/adriantoine/enzyme-to-json

import { TodoList } from '../../src/components/TodoList';
import * as todoStubs from '../../stubs/todos';

describe('TodoList', () => {
    let wrapper;
    beforeEach(() => {
        wrapper = shallow(<TodoList todos={todoStubs.sortedTodos}/>)
    });

    it('renders correctly', () => {
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});