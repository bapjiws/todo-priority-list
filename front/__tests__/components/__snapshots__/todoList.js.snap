// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList renders correctly 1`] = `
<Table
  bordered={true}
  bsClass="table"
  className="todos-container"
  condensed={true}
  hover={false}
  responsive={false}
  striped={true}
>
  <thead>
    <tr>
      <th>
        Name
      </th>
      <th>
        Description
      </th>
      <th>
        Priority
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        eat something
      </td>
      <td>
        ASAP
      </td>
      <td>
        4
      </td>
    </tr>
    <tr>
      <td>
        get some sleep
      </td>
      <td>
        ASAP, too
      </td>
      <td>
        3
      </td>
    </tr>
    <tr>
      <td>
        do the dishes
      </td>
      <td>
        not so urgent
      </td>
      <td>
        2
      </td>
    </tr>
    <tr>
      <td>
        do some work
      </td>
      <td>
        meh
      </td>
      <td>
        1
      </td>
    </tr>
  </tbody>
</Table>
`;
